### DISCUSSION TITLE
AltGr layer for diacritics

### DISCUSSION THREAD

**Discussion by NickG13** [binarybottle/engram-es#28] [Ideas]
I was talking with Lobo-Feroz and it emerged an idea which I applied to my personal layout: using third layer of certain letters as an alternative for less frequent (but related) characters. 

For example, I've put: 
- In the U key, the third layer could be ü 
- In the N key, the third layer could be ñ 
- In the C key, the third layer could be ç

**Comment by Lobo-Feroz** [comment]
I think this is an awesome idea just for the ü. 

This makes much more sense than having a special diacritic key just for this symbol, since the diaeresis is not used in any other circumstance.

I'd keep ( ´ + n = ñ ) just because I think it's easier to type, but we can have both versions.

**Comment by iandoug** [comment]
That is the approach I took to the trial layouts I made.

I think the previous suggestion was to put the magic diacritic key (MDK) on the AltGr layer. 
But then every diacritic letter requires three keys:
AltGr + MDK + letter
as opposed to, for example, putting it on the letter's AltGr layer, for example
AltGr + letter.

For Spanish, this works fine, except for the rare ü, which we put on the next closest letter instead, since u already has ú.

I did try some experiments replacing k and w with ó but you immediately run into SFB problems, from mixing vowels and consonents.

So Qwerty:
![qwerty es iso](https://user-images.githubusercontent.com/12673886/129470619-a0df2e88-bff7-4e01-ad43-26b58b880070.png)

Early version Poqdea on ANSI (can optimise further)
![podtea-qp es ansi](https://user-images.githubusercontent.com/12673886/129470738-a85f0296-e952-469d-92e2-25e9f7737ed0.png)

If we don't need the diacritic diaeresis then we can remove it and use ASCII " as combining character to make French letters. Same with acute accent.

**Reply by binarybottle** [reply]
Next nearest key to "u" is a good idea for ü, or in case new layouts arise, choosing a mnemonic key, such as : or "with two components.

How exactly are you proposing we make French letters?  It would be nice to be able to construct any letter with an acute or grave accent, or circumflex, or umlaut.

**Reply by iandoug** [reply]
I'm on Linux not Mac, we have the Compose Key function which can do all sorts of magic. 

https://en.wikipedia.org/wiki/Compose_key

There are software solutions for Win and Mac (Karabiner)

For example, Compose + letter + ' gives acute version ... é á í ĺ ś (assuming it is defined in config file)
Compose + letter + " gives ë ä ü ÿ etc
Compose + letter + ` gives è à ì ò 
Compose + letter + ^ gives ê â î û etc ... 
or ç or ñ or ł or Ł or £ or ° or ∞ or π or φ or even ☭ if needed :-) (that's Compose + CCCP)

Usually config file defines both letter + modifier and modifier + letter, so you could go Compose + ' + a instead to get á

French + Spanish + German quickly use up all AltGr spots. The Compose key is elegant solution (same config file also defines any Dead Key patterns) and is much superior to Windows "just memorise this list of numbers to type special characters" method.

**Reply by iandoug** [reply]
> How exactly are you proposing we make French letters?

I think there is a reason why French have their own layouts ... The Germans tried to cover multiple languages and it quickly becomes difficult.

https://en.wikipedia.org/wiki/QWERTZ#/media/File:German-Keyboard-Layout-T2-Version1-large.png

Their red X functions as a Compose key.

**Reply by binarybottle** [reply]
> > How exactly are you proposing we make French letters?
> 
> I think there is a reason why French have their own layouts ... The Germans tried to cover multiple languages and it quickly becomes difficult.
> 
> https://en.wikipedia.org/wiki/QWERTZ#/media/File:German-Keyboard-Layout-T2-Version1-large.png
> 
> Their red X functions as a Compose key.

What a mess!

**Reply by binarybottle** [reply]
> I'm on Linux not Mac, we have the Compose Key function which can do all sorts of magic.
> 
> https://en.wikipedia.org/wiki/Compose_key
> 
> There are software solutions for Win and Mac (Karabiner)
> 
> For example, Compose + letter + ' gives acute version ... é á í ĺ ś (assuming it is defined in config file)
> Compose + letter + " gives ë ä ü ÿ etc
> Compose + letter + ` gives è à ì ò
> Compose + letter + ^ gives ê â î û etc ...
> or ç or ñ or ł or Ł or £ or ° or ∞ or π or φ or even ☭ if needed :-) (that's Compose + CCCP)

Thank you -- I love that last one!  So these can be typed sequentially?

**Comment by binarybottle** [comment]
I re-read all the comments and suggestions from @iandoug, @Lobo-Feroz, and @NickG13 (thank you, team!) and came up with the following. I don't know what's practical with respect to using a compose key across platforms, so I incorporated the accent / AltGr layers for diacriticals.  Please tell me what you think.
![PXL_20210822_040207101](https://user-images.githubusercontent.com/236532/130342123-8527d916-b8bc-44bd-856a-1a2127d8ab35.jpg)

**Reply by binarybottle** [reply]
Note that I relegated the single and double quotation marks to the ISO key on the bottom left since " is replaced with << and >> and ' is rarely used, as I understand.

**Reply by binarybottle** [reply]
Would it be better to swap << with < and >> with >, since quotations are more common than use of the greater-than and less-than signs?  Would accessing < and > with AltGr be a problem?

**Reply by iandoug** [reply]
![blank-iso-main](https://user-images.githubusercontent.com/12673886/130346053-845399c2-e9ff-4eb6-a0cb-4e9b3fceaa52.jpg)

**Reply by iandoug** [reply]
How does the output of your generator look? May be able to knock together a raw-output-to-pretty-picture.

**Reply by Lobo-Feroz** [reply]
_**Important edit:** while I liked it on first sight, testing it in type-fu made me realize that this current placement of the diacritical key is **very uncomfortable**. I have a comment below with a type-fu layout explaining this._

I like a lot this one. By moving the parenthesis to the bracket keys, you have more freedom for the symbol center columns, no need to juggle anymore the shifts on the upper row, or the exclamation signs on the lower row. Moving the brackets to lower layers is not a problem since they're not that frequent. This is brilliant.

One thing with the diacritics key in the place of qwerty's "Y" is that the combo AltGr+Y is not easy on the fingers. Try it in your keyboard. You have to tuck the thumb under your palm then stretch the index to reach the Y key. It's a very weird position. For spanish it's not a big deal since we dont use those two diacritics ´ ^. For french it might be more problematic. At this moment I don't know any french-speaking person that could enlighten us.

Since this seems a powerful first version to implement, let me mention some minor details:

- We would need an alternative for the quotes key on ANSI keyboards. This can be an AltGr of something somewhere.
- We have the € in AltGr+€. Another idea could be to have $ in AltGr+S (not that there is any problem with having them also in the # key).
- The middle dot · could be AltGr + period.
- The dash — could be AltGr + hyphen.

I'll set up now a type-fu version of this layout to try it out for real.

**Reply by Lobo-Feroz** [reply]
> Would it be better to swap << with < and >> with >, since quotations are more common than use of the greater-than and less-than signs? Would accessing < and > with AltGr be a problem?

I don't really know what to tell you about this, for this reason: normal people in daily life don't use spanish quotes because, as you know, they are not on the qwerty-es keyboard, but I don't know if they would use them more if they were available.

What I can tell you is that <> are currently on their own key  (base: < , shift: > ) and we're used to type them rather quickly. 

We could have a first version as in the drawing, and then collect additional feedback from the users.

Another thing I wanted to mention is the asterisk *. It seems is frequent enough, much more in the Gutembert corpus because there they use it for emphasizing titles, but even in the Leipzig corpus where it is less used, it still ranks fairly above other symbols like @ or #. I propose that one of the keys next to enter could go to the asterisk. We could have an "internet key" with # and @, and the asterisk in the other one. Or group * with @ since both are round(ish).

**Reply by iandoug** [reply]
> . For french it might be more problematic. At this moment I don't know any french-speaking person that could enlighten us.

https://github.com/phques
May be able to help if I could figure out how to mail him ...
Let's try this: @phques

**Reply by Lobo-Feroz** [reply]
OK I must stand corrected.

I made the typefu of this version and tested it. Layout file and screenshot:

[Engram-ES - V0 - 20210822.tfl.zip](https://github.com/binarybottle/engram-es/files/7027331/Engram-ES.-.V0.-.20210822.tfl.zip)

![imagen](https://user-images.githubusercontent.com/10242383/130355471-e51db497-487d-45b7-92ab-6fca0f1f45c8.png)

The big problem: the current diacritic key in center right column, upper row is **_very uncomfortable_**. I have tested it and in a staggered keyboard moving the right index finger from its home position (qwerty J, engram-es T) to the diacritics position (qwerty Y), is a very hard movement, I can't accomplish it with my finger alone, I need to move my whole hand (and I have relatively big hands).

It's for sure not a movement I want to make several times per sentence, since there are usually accented vowels every few words.

Heck, it's even harder than doing it in the current qwerty-es, since the current accent key is one key to the left of the right pinky (where the K is here in engram-es), and I can hit it moving just my pinky one step to the right.

I can't seem to make a list of words or sentences with tildes on type-fu (it doesn't seem to recognize them as valid), if I manage to make it I'll upload it here.

Note: conversely, moving the left index to the center bottom row (qwerty B) is harder due to the staggering. But we don't use ¿¡ that much: at most, once per exclamation sentence. Qwerty T and N, which would be engram-es -_ and ?! , are easier, again due to the staggering. Of course in ortholinears this would be different. The main problem is with the ´ in the Y key, which would be an uncomfortable movement, and used _a lot_.

**Reply by binarybottle** [reply]
> > . For french it might be more problematic. At this moment I don't know any french-speaking person that could enlighten us.
> 
> https://github.com/phques
> May be able to help if I could figure out how to mail him ...
> Let's try this: @phques

I work with a group in Paris, so I can reach out to them.

**Reply by phques** [reply]
Hi, what would you like to know?

**Reply by iandoug** [reply]
> Hi, what would you like to know?

Thanks :-)
See https://github.com/binarybottle/engram-es/discussions/28#discussioncomment-1218418
This is layout for Spanish, but they're trying to accommodate French diacritics as well.. and are looking for suitable inputs :-)

**Reply by phques** [reply]
> 
> One thing with the diacritics key in the place of qwerty's "Y" is that the combo AltGr+Y is not easy on the fingers. Try it in your keyboard. You have to tuck the thumb under your palm then stretch the index to reach the Y key. It's a very weird position. For spanish it's not a big deal since we dont use those two diacritics ´ ^. For french it might be more problematic. At this moment I don't know any french-speaking person that could enlighten us.
> 
You can have a look here for a letter frequency chart on Wikipedia, the corpus is the French Wikipedia (2008).
(the ones in green are NOT valid French, ignore them)
  https://fr.wikipedia.org/wiki/Fr%C3%A9quence_d%27apparition_des_lettres_en_fran%C3%A7ais
'é' is very common, the Canadian French layout has a key assigned for it.
'à' is the word "at" .. so pretty common too ;-) 
The MacOS Canadian French layout has an interesting way of handling this : ùàèéç all have an assigned key and have other characters on Alt-Gr  (Option). Of course this is only makes sense if you're creating a French layout !! 
Personally, when using my own (English) layouts, I found that trying to have support for French was never satisfactory and I end up just switching to Canadian French to write in French.

**Comment by binarybottle** [comment]
Thank you for your prompt feedback, @Lobo-Feroz!  Please tell me what you think about the following:

![PXL_20210822_162210918](https://user-images.githubusercontent.com/236532/130362707-f2e50da7-ed2c-4acc-9061-2d68927def3f.jpg)

**Reply by Lobo-Feroz** [reply]
Thanks for the quick fix @binarybottle . 

Type-fu layout and screenshot:

[Engram-ES - V0 - 20210822-02 - diacritics home row.tfl.zip](https://github.com/binarybottle/engram-es/files/7027789/Engram-ES.-.V0.-.20210822-02.-.diacritics.home.row.tfl.zip)

![imagen](https://user-images.githubusercontent.com/10242383/130363844-bdbf0639-425d-4c1b-b9dc-f31c6fd84e05.png)

This one is much easier on the diacritics than the previous one.

Two little things:

- Please consider switching the places of the hyphen and the asterisk, for two reasons: first, I think the ISO key is easier than the next-to-enter key, it's just moving the left pinky straight down. Since the hyphen is more frequent than the asterisk I think it should go in the easier key. Second, though this actually should not matter when designing an optimal keyboard, the next-to-enter key already has the asterisk in qwerty-es ([qwerty-es layout](https://upload.wikimedia.org/wikipedia/commons/thumb/7/74/KB_Spanish.svg/400px-KB_Spanish.svg.png)). Here it's the shift layer to the base +, but it's clear than the asterisk is more frequent than the plus.

- Since we're adding currency to the AltGr layer, I think we could have the cent in the c. There are many spanish speakers in the USA and I think that symbol is used there - please correct me if I'm wrong. (Also we have cents of euro too, though we don't use that symbol, we write 0.50 € instead).

Still unable to add words with diacriticals to a type-fu practice text. Will write the dev to see if he can help.

**Reply by binarybottle** [reply]
Thanks, @Lobo-Feroz!

I have one problem with switching the places of the hyphen and the asterisk: I want to support ANSI keyboards, so that ISO key needs to be expendable. The way I have it currently, the asterisk is a duplicate and is expendable, whereas the hyphen is not.  

I like the addition of the cent symbol, but in my experience it is rarely used.  No harm in including it, is there?

**Reply by Lobo-Feroz** [reply]
> I have one problem with switching the places of the hyphen and the asterisk: I want to support ANSI keyboards, so that ISO key needs to be expendable. The way I have it currently, the asterisk is a duplicate and is expendable, whereas the hyphen is not.

Thought about that. Is it possible to make it that in the ANSI keyboard the key removed would be the * instead? 

In other words, when you set up a keyboard, say, for windows, you have to do two different layouts one for ANSI and other for ISO, or it's one layout file with the possibility of it being used in both ANSI and ISO hardware? 

In the MS Keyboard Layout Creator, I have an option choice between three shapes of keys around the enter. One looks like ISO and another like ANSI. But every choice still shows me the ISO key on the bottom row. I don't seem to have the option of choosing a strict ANSI keyboard.

If setting slightly different ISO and ANSI layouts is not possible, then of course we have to place there an expendable character.

**Reply by iandoug** [reply]
@Arno... I still think D should be home key, not T.

**Reply by iandoug** [reply]
There are at least 3 styles
I have an ANSI cheap keyboard that looks like the ISO105, except key next to Z is removed and replaced with large shift.

Laptops are a whole different minefield.

![ansi104](https://user-images.githubusercontent.com/12673886/130366525-351c61b9-2a2f-4c6f-a5b2-ce37709e0cf5.jpg)
![ansi104bae](https://user-images.githubusercontent.com/12673886/130366533-624ddc41-d064-4ee4-b805-bef7ec2f1921.jpg)
![iso105](https://user-images.githubusercontent.com/12673886/130366535-b23bd79e-01dc-410a-bdd2-db779fafe201.jpg)

**Reply by iandoug** [reply]
> In the MS Keyboard Layout Creator, I have an option choice between three shapes of keys around the enter. One looks like ISO and another like ANSI. But every choice still shows me the ISO key on the bottom row. I don't seem to have the option of choosing a strict ANSI keyboard.

Linux text-only config allows separate definitions. Extract from GB/UK file, this is a patch to extned US ANSI Coelmak for ISO 105 UK boards:
```
// Describe the differences between the US Colemak layout
// and a UK variant. By Andy Buckley (andy@insectnation.org)

partial alphanumeric_keys
xkb_symbols "colemak" {
    include "us(colemak)"

    name[Group1]="English (UK, Colemak)";

    key <AE02>	{ [         2,   quotedbl,  twosuperior,    oneeighth ]	};
    key <AE03>	{ [         3,   sterling, threesuperior,    sterling ]	};
    key <AE04>	{ [         4,     dollar,     EuroSign,   onequarter ]	};

    key <AC11>	{ [apostrophe,         at, dead_circumflex, dead_caron]	};
    key <TLDE>	{ [     grave,    notsign,          bar,          bar ]	};

    key <BKSL>	{ [numbersign, asciitilde,   dead_grave,   dead_breve ]	};
    key <LSGT>	{ [ backslash,        bar,          bar,    brokenbar ]	};

    include "level3(ralt_switch)"
};
```

**Reply by binarybottle** [reply]
> @arno... I still think D should be home key, not T.

@iandoug -- While D and C are more frequent than T, the bigrams that contain T and S, N, or R have much higher frequencies  than the bigrams that contain D and S, N, or R -- here are the top ones:

> NT    13403852
> ST    10479844
> TR   6982844
> ND     6501402
> PR     5241844
> NC     4535985
> RT     3473930
> ...
> 

And none of the top layouts that the optimization algorithm generated included D on the home row:

> Rank                                                   Score
> 1:  Z H O X P I E A F Y Q U M D B J T S N R C L V G    0.718175779085395
> 1:  Y H O X P I E A F Z Q U M D B J T S N R C L V G    0.7181740824015593
> 1:  Z H O X P I E A F Y Q U M C B J T S N R D L V G    0.7181662126172352
> 1:  Y H O X P I E A F Z Q U M C B J T S N R D L V G    0.7181645159333996
> 1:  Z H O X P I E A F Y Q U M L B J T S N R C D V G    0.7181440154199303
> 1:  Y H O X P I E A F Z Q U M L B J T S N R C D V G    0.7181423187360948
> 1:  Z H O X P I E A F Y Q U M L B J T S N R D C V G    0.7181399906496914
> 1:  Y H O X P I E A F Z Q U M L B J T S N R D C V G    0.7181382939658557
> 1:  Z H O X P I E A F Y Q U M D B J C S N R T L V G    0.7181368885518766
> 1:  Y H O X P I E A F Z Q U M D B J C S N R T L V G    0.7181351918680411
> 1:  Y H O X P I E A Z F Q U M D B J T S N R C L V G    0.7181315366108767
> 1:  F H O X P I E A Z Y Q U M D B J T S N R C L V G    0.7181279794445848
> 
> Layout 1:
> 
> Z H O X  M D B J
> P I E A  T S N R
> F Y Q U  C L V G
>

**Reply by iandoug** [reply]
I can turn those layout strings into KLA json and then into images, but 
1. Punctuation?
2. Are K and W not relevant to Algo?

Trying to understand why you ignore DE, DO, AD, or alternatively, why appearance in bigram is more important than appearance alone...

Thanks, Ian

**Reply by iandoug** [reply]
Attached
[json.zip](https://github.com/binarybottle/engram-es/files/7031018/json.zip)
[img.zip](https://github.com/binarybottle/engram-es/files/7031021/img.zip)

**Reply by binarybottle** [reply]
> Trying to understand why you ignore DE, DO, AD, or alternatively, why appearance in bigram is more important than appearance alone...
> 

If you read the documentation in the Jupyter notebook (or in the README, but this needs to be updated), you will see that all letter pairs are accounted for -- I don't mention the D+vowels because they are on opposite hands.  Letters are accounted for in order of decreasing frequency.

**Reply by binarybottle** [reply]
> Attached
> [json.zip](https://github.com/binarybottle/engram-es/files/7031018/json.zip)
> [img.zip](https://github.com/binarybottle/engram-es/files/7031021/img.zip)

These look great!  Can't wait till they are complete!

**Reply by Lobo-Feroz** [reply]
Type-fu 5.3.2 supports correctly the diacritics in the dead key (thanks to the author Jarek Foksa for releasing this fix).

(If it doesn't update you need to "relaunch" the app from the main url type-fu.com)

I've created a words dictionary with the 10K most common words in spanish based on @iandoug 's "goodwords" file. It needs some cleanup for duplicates but for the purposes of testing the keyboard it works perfectly fine:

[Spanish words - 10K - 20210824.tfd.zip](https://github.com/binarybottle/engram-es/files/7035443/Spanish.words.-.10K.-.20210824.tfd.zip)

With both the layout and this sample text file anyone who wants to test engram-es can do it via this website. Now moving on to the windows layout...

**Reply by Lobo-Feroz** [reply]
> There are at least 3 styles
> I have an ANSI cheap keyboard that looks like the ISO105, except key next to Z is removed and replaced with large shift.

Thanks for the reference @iandoug. I'll be setting up the windows layout next, I don't know how complex would it be to make a slightly different layout for ANSI/ISO or if possible at all, so I'll focus on creating the current one that has the asterisk on the ISO key, for starters.

(I'm still curious about this, so I'll try to look it up and make another thread for this ANSI/ISO issue, but I really want to have a working engram-es keyboard ASAP)

**Comment by binarybottle** [comment]
@iandoug, @Lobo-Feroz, @NickG13 -- I put the latest layout in the README as a placemarker until we finalize our diacritical strategy: https://github.com/binarybottle/engram-es/blob/master/README.md

**Reply by Lobo-Feroz** [reply]
Thanks for the mentions Arno ! I think this version is good enough to make public and see if other people can give more usage feedback. I'm having a go at type-fu now but next thing I want to do is the windows layout.

A copypaste thingy: "AltGr + 6 = » (open quote/comillas)" should read "close quote"

**Reply by binarybottle** [reply]
Thank you -- fixing now!

**Comment by iandoug** [comment]
☆ + Shift         + [letter] = [letter] with a diaresis/umlaut: ü
☆         + AltGr + [letter] = [letter] with a grave accent: è
☆ + Shift + AltGr + [letter] = [letter] with a circumflex: â

Um, how do you do the Capital versions?

Mockup attached. I added single and double typographic quotes, ordinal indicators, and Not sign.

Have not attempted to add the diacriticed versions.

![engram es ansi](https://user-images.githubusercontent.com/12673886/130573383-b1bd27e0-3641-4d3b-9b03-e2faa07c20f3.jpg)

[engram.es.ansi.json.txt](https://github.com/binarybottle/engram-es/files/7037215/engram.es.ansi.json.txt)

**Reply by Lobo-Feroz** [reply]
> Um, how do you do the Capital versions?

It's a dead key, so two strokes. Capital character is shift in the second stroke.

Note however, as I've said in a previous comment, that this seems very uncomfortable for the AltGr diacritics è and ê. Spanish does not use them, but other languages might require other solutions.

Spanish only uses áéíóú and ü, and for those the key function is the same as before, it's only physically displaced (before, it was in the place of the K here, next to the enter).

**Reply by Lobo-Feroz** [reply]
By the way, how do you upload a json to keyboard-design.com? I managed to export one, but couldn't find the way to import.

**Reply by iandoug** [reply]
> By the way, how do you upload a json to keyboard-design.com? I managed to export one, but couldn't find the way to import.

do you mean klanext.keyboard-design.com ?
On the Configure page, click Import button, then paste the Json into the dialogue and click Import.

Am looking at making conventional upload/download operations work.

**Reply by iandoug** [reply]
Be aware that KLA has no concept of dead keys (yet), and if the required letter is not on the keyboard, it is ignored, which affects the score.

Layouts that DO have the letter "type" it, pushing up their scores.

So at the moment we can't fairly use KLA to evaluate layouts.

If you want to add a layout permanently to KLAnext, email it to me.
